### PR TITLE
feat: add strategy mocks

### LINKED ISSUE #25: Create BaseStrategy and tests
Implement BaseStrategy (Solidity)
------
**Description:** 
Create BaseStrategy contract so that writing a strategy is as simple as possible. 

BaseStrategy is an abstract contract that is inherited by actual strategies. 

Use BaseStrategy from the v2 rewrite in solidity: https://github.com/saltyfacu/vaults-hardhat/blob/develop/contracts/core/BaseStrategy.sol

This version should leave Access Control references as: 
`# TODO: permissioned: ROLE
`to indicate that, when the AccessControlRegistry contract is created, strategies should reference to it

Create several TestStrategyFLAVOR.sol file in test folder that implements the simplest possible instance of a strategy (by just moving tokens) so it can be used during tests. 

Flavors of strategies to test: 
- regular liquid strategy
- strategy that locks liquidity (not possible to withdraw until some time passes)
- strategy that losses money when withdrawing only (similar to slippage when getting out, or withdrawal fee) 

**Dependencies:**
NA

**Acceptance criteria:** 
- contracts/BaseStrategy.sol is created
- contracts/test/TestStrategyLiquid.sol is created and tested
- contracts/test/TestStrategyLocked.sol is created and tested
- contracts/test/TestStrategyLossy.sol is created and tested

**Payment Trigger:** 
PR is merged
**Category:**
DEV
**Level:**
1

### LINKED ISSUE #31: Implement updateDebt + tests
Description:
- implement `updateDebt` in the vault and write corresponding tests. 

`updateDebt` should handle how the vault manages the debt of the strategy. If there is room to increase the debt, then it should push funds from the vault into the strategy. If the debt the strategy can handle has decreased, it should pull funds from the strategy. 

- write tests to handle `updateDebt` as well. 

Dependencies:
N/A

Acceptance criteria:
- implement updateDebt
- implement unit tests associated with updateDebt

Payment Trigger:
PR is merged
Category:
<>
Level:
<>

### LINKED ISSUE #32: Implement `processReport` + tests
Description:
- implement `processReport` in the vault and write corresponding tests. 

`processReport` should handle strategy updates to the vault regarding any gains or losses it incurs. the vault should update its own registry of the strategy's debt, gains and losses. 

- write tests to handle `processReport` as well. 

Dependencies:
N/A

Excluded:
- healthcheck is not implemented or handled
- feemanager module is not implemented or handled

Acceptance criteria:
- implement processReport
- implement unit tests associated with processReport

Payment Trigger:
PR is merged
Category:
<>
Level:
<>

### PR DESCRIPTION
**Pull request by anticlimactic**
## Description

Added basic strategy mocks for the three types: liquid, locked and lossy. 
I assumed `withdrawable` informs the vault of how many assets are left.

I considered adding a way of accounting for fees (for example, slippage or withdrawal fees from the protocols we use, via `freeFunds`) but might leave that until we have a proper way of adding and remove debt to the strategy from the vault. 

Closes #25 
Closes #31 
Closes #32 

## Checklist

- [x] I have run vyper and solidity linting
- [x] I have run the tests on my machine
- [x] I have followed commitlint guidelines
- [x] I have rebased my changes to the latest version of the main branch

### CODE REVIEW

File: contracts/test/mocks/LockedStrategy.sol [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    uint256 balance = IERC20(asset).balanceOf(address(this));
+    if (_amount > balance) revert InsufficientFunds();
+    lockedBalance = _amount;
+    lockedUntil = block.timestamp + _lockTime;
+  }
+
+  // only used during testing
+  // free locked funds if duration has passed
+  function freeLockedFunds() external {
+    uint256 balance = IERC20(asset).balanceOf(address(this));
```
Comment (jmonteer) [inline contracts/test/mocks/LockedStrategy.sol:30]: this line is not used

**Review by jmonteer — COMMENTED** [review]
We need to implement the interaction between vault and strategies for these tests to be completed. 

That way we can test that a strategy can be withdrawed when locked and everything.

File: contracts/VaultV3.vy [hunk 2]
```diff
@@ -292,11 +298,65 @@ def migrateStrategy(new_strategy: address, old_strategy: address):
...
     # TODO: should this emit an event?
 
+
+@external
+def updateDebt(strategy: address) -> uint256:
+    # TODO: permissioned: DEBT_MANAGER (or maybe open?)
+    # TODO: rebalance debt. if the strategy is allowed to take more debt and the strategy wants that debt, the vault will send more. if the strategy has too much debt, the vault will have less
+    currentDebt: uint256 = self.strategies[strategy].currentDebt
+
+    minDesiredDebt: uint256 = 0
```
Comment (pandadefi) [inline contracts/VaultV3.vy:313]: Is the initialization to zero value required?

Comment (anticlimactic) [reply contracts/VaultV3.vy:313]: apparently so. i get this error when i tried to declare without initial assignment

```
vyper.exceptions.VariableDeclarationException: Memory variables must be declared with an initial value
```

**Comment by anticlimactic** [comment]
believe the tests are failing due to the failing test in master. but this is ready for review.

**Review by jmonteer — APPROVED** [review]
Do we need a SlippageStrategy which incurs in losses while withdrawing? 

Maybe for another ticket

**Pull request merged**